File: lib/domain/field_cultivation/gateways/field_cultivation_climate_source_gateway.rb
# frozen_string_literal: true

module Domain
  module FieldCultivation
    module Gateways
      # Narrow read I/O for climate data (plan access + climate source + weather targets).
      class FieldCultivationClimateSourceGateway
        # @return [Domain::FieldCultivation::Dtos::FieldCultivationPlanAccessSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordNotFound]
        def find_plan_access_snapshot_by_field_cultivation_id(field_cultivation_id)
          raise NotImplementedError,
                "Subclasses must implement find_plan_access_snapshot_by_field_cultivation_id"
        end

        # @return [Domain::FieldCultivation::Dtos::FieldCultivationClimateSourceSnapshot]
        # @raise [Domain::Shared::Exceptions::RecordNotFound]
        def find_climate_source_snapshot_by_field_cultivation_id(field_cultivation_id)
          raise NotImplementedError,
                "Subclasses must implement find_climate_source_snapshot_by_field_cultivation_id"
        end

        # @return [Domain::WeatherData::Dtos::WeatherPredictionTargets]
        def find_weather_prediction_targets_by_plan_id(plan_id)
          raise NotImplementedError
        end
      end
    end
  end
end

File: crates/agrr-domain/src/field_cultivation/gateways/field_cultivation_climate_source_gateway.rs
use crate::field_cultivation::dtos::{
    FieldCultivationClimateSourceSnapshot, FieldCultivationPlanAccessSnapshot,
    WeatherPredictionTargets,
};

/// Ruby: `FieldCultivationClimateSourceGateway`
pub trait FieldCultivationClimateSourceGateway: Send + Sync {
    fn find_plan_access_snapshot_by_field_cultivation_id(
        &self,
        field_cultivation_id: i64,
    ) -> Result<FieldCultivationPlanAccessSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn find_climate_source_snapshot_by_field_cultivation_id(
        &self,
        field_cultivation_id: i64,
    ) -> Result<FieldCultivationClimateSourceSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn find_weather_prediction_targets_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<WeatherPredictionTargets, Box<dyn std::error::Error + Send + Sync>>;
}
